chore(KFLUXVNGD-171): add task to update and push trusted tasks

avi-biton opened the pull request:
Add a step to update and push the list of trusted tasks This task will be a part of a new release pipeline called `push-bundles-to-external-registry` that will release tekton tasks as bundles and will update the list of trusted-tasks

avi-biton committed:
chore(KFLUXVNGD-171): add step to update and push trusted tasks

Add a step to update and push the list of trusted tasks
This task will be a part of a new release pipeline called
`push-bundles-to-external-registry` that will release
tekton tasks as bundles and will update the list of
trusted-tasks
Add tests

Signed-off-by: Avi Biton <abiton@redhat.com>

avi-biton commented:
@konflux-ci/release-service-maintainers can I please get a review?

avi-biton commented:
/retest

davidmogar commented:
/retest

johnbieren commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+metadata:
+  name: update-trusted-tasks
+  labels:
+    app.kubernetes.io/version: "1.9.0"
```
This should be 0.1.0

johnbieren commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+        do
+            component=$(jq -c --argjson i "$i" '.components[$i]' "${SNAPSHOT_SPEC_FILE}")
+            repository=$(jq -r '.repository' <<< "$component")
+            imageTag=$(jq '.tags' <<< "$component" | jq -r '.[0]')
```
It looks to me (based on the unit tests) that you expect this to return `latest` if a `latest` tag exists. I see `latest` is the first one in your Tags list in the mocks. Will anything break if the tags are like `["a", "b", "latest", "v4.5"]`? Essentially `latest` exists but isn't `[0]`

avi-biton replied:
There are 2 OCI artifacts we are handling in this task:
1 - the task bundles that we push and release, and that `imageTag` is its tag as defined in the Snapshot, we are taking the first tag.
2 - After releasing the bundle, we also update an OCI artifact called `data-acceptable-bundles` with the new task that we just added.

The `ec track bundle` command depends on whether `data-acceptable-bundles:latest` already exist or not.
Thats what we check with `skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep ${TAG} &> /dev/null` 

Anyways, I modified the tags return values in the mock to be `'{"Tags": ["v2.0.0-3", "latest", "v2.0.0-2"]}'` just to demonstrate that the order doe not make a difference

mmalina replied:
I would suggest:
```
imageTag=$(jq '.tags[0]' <<< "$component")
```

But also, I'm not sure why the first tag is used. Why not use the digest sha reference directly?

avi-biton replied:
When using the `ec track bundle --freshen` , the bundle we are adding is referred to by its tag:
`ec track bundle --freshen --bundle "${repository}:${imageTag}" --output oci:"${ACCEPTABLE_BUNDLES}:${TAG}"`

mmalina replied:
Are you saying sha reference would not work here?

avi-biton replied:
We are using the `--freshen` option which resolves the image tag to use the latest image of this tag
https://conforma.dev/docs/ec-cli/ec_track_bundle.html#_options
Thats the reason we are using the tag and not the sha reference.

mmalina replied:
So you actually intentionally want the latest even if some other image was pushed to that tag in the meantime? That seems a bit odd, normally you want to be sure that you're releasing the thing that's in the snapshot. But I don't fully understand how this process works, so maybe it's ok.

avi-biton replied:
Thanks a lot for the comments, you made me rethink my approach.
I have updated the task to get the data from the `push-snapshot-results.json` and update the trusted-tasks according to the images in this file and to run the ec command without `--freshen` so that I would update the sha as it appears in the `push-snapshot-results.json`.

I will also update the tests accordingly.

avi-biton replied:
New approach and tests were implemented

mmalina replied:
OK, cool. I was gonna ask why not use the snapshot spec rather than the push-snapshot results, but I can see you're using the urls from there, which I think is basically all the tags, right? Can you explain why you need to add each tag as a trusted bundle? Is it so that people can reference them via any of the tags? And even so, couldn't you just use the tags from the mapped snapshot spec? It will have all the tags in each component. Maybe it's ok to use the results file, but I don't think we ever relied on those for anything, so I'm not sure if we want to start doing that now - especially if you can get the same data from the snapshot spec.

avi-biton replied:
I get the same data from the snapshot spec, but I think that using the push-snapshot results is a better approach since we get the information of the actual pushed images and use it to update the trusted-tasks.
Is there a specific reason not to use it?

avi-biton replied:
I am updating each tag because I want to support multiple tags in the RPA

mmalina replied:
> Is there a specific reason not to use it?

The only reason I can think of is consistency. Because I don't think we use it anywhere else. The results data is stored in some tasks and then the `update-cr-status` task will inject that data into the Release CR. For example, if we decide to change the format in the future, there is a potential of breakage if we don't realize that some other task is relying on this data.

You're right that it sounds better that here you know for sure that this is what got pushed. But on the other hand, if something that's in the mapped snapshot is not pushed and the pipeline still passes, we have a bug. So it should be ok to just use it from the snapshot.

I would prefer to use the snaphost, but I don't have a strict opinion here. So maybe @johnbieren can chime in?

avi-biton replied:
OK, I understand the risk you mentioned.
I can revert it back to use the snapshot spec if you think this approach is better (mainly because I don't want to break it in case the results format will be updated).

johnbieren replied:
I think the snapshot spec is better because we have a schema for it and it is more well defined IMO. I am not even sure of the structure of the results file off the top of my head (I know I could check it, but). I also could foresee a world where we say "oh hey the results are a lot prettier in the UI if we format it like this instead" and totally forget about this task. The snapshot spec is more internal (so we wouldn't do a "pretty" update to it) and a change to it would affect *so* many tasks so we would be more careful to update them all and it is less likely this one would be left out

avi-biton replied:
Gotcha, thanks a lot

johnbieren reviewed (dismissed):
Missing a README.md

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+  params:
+    - name: snapshotPath
+      type: string
+      description: Path to the JSON string of the Snapshot spec in the config workspace to apply the mapping to
```
> in the config workspace

But there is no config workspace, it's called `data`. Also, to apply what mapping? It seems it's copied from apply-mapping or something.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+      description: Path to the JSON string of the Snapshot spec in the config workspace to apply the mapping to
+  workspaces:
+    - name: data
+      description: The workspace where the extra config file containing the mapping and snapshot json reside
```
This should be updated as well. It seems you're only using it for the snapshot.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+            component=$(jq -c --argjson i "$i" '.components[$i]' "${SNAPSHOT_SPEC_FILE}")
+            repository=$(jq -r '.repository' <<< "$component")
+            imageTag=$(jq '.tags' <<< "$component" | jq -r '.[0]')
+            ACCEPTABLE_BUNDLES=$(echo "${repository}" | awk -F'/' '{$NF="data-acceptable-bundles"; print $0}' OFS='/')
```
It's the containerImage pull spec for that? I think it would be better to fetch it using `jq` rather than parsing a json manually like this.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+            imageTag=$(jq '.tags' <<< "$component" | jq -r '.[0]')
+            ACCEPTABLE_BUNDLES=$(echo "${repository}" | awk -F'/' '{$NF="data-acceptable-bundles"; print $0}' OFS='/')
+            set +e
+            skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep ${TAG} &> /dev/null
```
How about doing `skopeo inspect docker://"${ACCEPTABLE_BUNDLES}:latest"` directly?

avi-biton replied:
I tried using skopeo inspect, but it is unsupported for this OCI artifact.
```
$ skopeo inspect docker://quay.io/avi_test/user-ns2/release-catalog/data-acceptable-bundles:latest
FATA[0001] unsupported image-specific operation on artifact with type "application/vnd.unknown.config.v1+json"
```

I then switched to skopeo list-tags and checking if `latest` exists

mmalina replied:
I see, I didn't realize this was an oci artifact. Then `oras resolve`? We use it for the same purpose here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/managed/push-snapshot/push-snapshot.yaml#L61

My worry is that listing all tags can be much heavier. But maybe in this case you only use latest, so probably not really a concern? So I leave it up to you.

avi-biton replied:
I understand your concern, but as you mentioned we only use latest.
I think we will keep it as it is now, and in time, if we will see it causes issues (unlikely) we can switch to oras

mmalina replied:
ok, sounds good.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,57 @@
+            component=$(jq -c --argjson i "$i" '.components[$i]' "${SNAPSHOT_SPEC_FILE}")
+            repository=$(jq -r '.repository' <<< "$component")
+            imageTag=$(jq '.tags' <<< "$component" | jq -r '.[0]')
+            ACCEPTABLE_BUNDLES=$(echo "${repository}" | awk -F'/' '{$NF="data-acceptable-bundles"; print $0}' OFS='/')
```
I had a hard time figuring out what this was. Maybe add a comment saying that this replaces the repo with "data-acceptable-bundles", e.g. quay.io/myorg/myrepo will be quay.io/myorg/data-acceptable-bundles.

avi-biton replied:
I will add a comment to explain it

avi-biton replied:
@mmalina I updated the description and added comment, I hope its clearer now

mmalina replied:
Thanks.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,82 @@
+    updated with the new task.
+    If an OCI artifact named data-acceptable-bundles is already in place we will use
+    it as an input and append it, else we will create a new artifact
+
```
Empty line - should probably be removed.

avi-biton replied:
Empty line removed

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,82 @@
+            # Get the first imagTag of the component  
+            imageTag=$(jq '.tags' <<< "$component" | jq -r '.[0]')
+
+            # The data-acceptable-bundles repo should reside in the same registry
```
In the same org in the registry, right? Not just the same registry.

avi-biton replied:
right, I updated it :)

johnbieren commented:
You're still missing a README

avi-biton commented:
> You're still missing a README
README was added

yftacherzog commented on tasks/managed/update-trusted-tasks/README.md:
```diff
@@ -0,0 +1,12 @@
+Tekton task to update trusted-tasks list bundles.
+Once a task is released as a bundle, the list of trusted-tasks should be 
+updated with the new task.
+The OCI artifact containing the trusted-tasks list named `acceptable-data-bundles` and should reside in the same org in the registry.
```
```suggestion
The OCI artifact containing the trusted-tasks list is named `acceptable-data-bundles` and should reside in the same org in the registry.
```

yftacherzog commented on tasks/managed/update-trusted-tasks/README.md:
```diff
@@ -0,0 +1,12 @@
+Once a task is released as a bundle, the list of trusted-tasks should be 
+updated with the new task.
+The OCI artifact containing the trusted-tasks list named `acceptable-data-bundles` and should reside in the same org in the registry.
+If it is already in place it will be used as an input and append to it, else a new artifact will be created
```
```suggestion
If it is already in place, it will be used as an input to which the results will be appended, else a new artifact will be created.
```

avi-biton commented:
@johnbieren @mmalina Thanks a lot for the reviews.
I updated the script to get the data from the snapshot spec file, and also to support multiple tags.
I also updated the tests and the README file.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,91 @@
+                
+                set +e
+                # Check if ACCEPTABLE_BUNDLES OCI artifact has a latest tag
+                skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep ${TAG} &> /dev/null
```
`ACCEPTABLE_BUNDLES` is set once per component. It seems like a waste to do this same request repeatedly for each tag. It should be done once before this for loop and the output saved to a tmp file and here you would just do the grep. But also, you should be careful about partial matches - the grep should use something like `"^${TAG}$"`.

avi-biton replied:
I was considering the following scenario:
Component with multiple tags with the same `ACCEPTABLE_BUNDLES` url and `ACCEPTABLE_BUNDLES`  does not exist.
For the first tag skopeo will return error (the tag does not exist), and will run the `ec track bundle` command without the input parameter. 
However, for the second tag in the same component, skopeo will return 0 because we already created the `ACCEPTABLE_BUNDLES` in the previous tag, and therefor will use the `ec track bundle` command with the input parameter. 
In addition, I was thinking that most of the components will only have one tag.

avi-biton replied:
The `ACCEPTABLE_BUNDLES` is set only once per component and not for each tag:
At line 59:
`ACCEPTABLE_BUNDLES=$(echo "${repository}" | awk -F'/' '{$NF="data-acceptable-bundles"; print $0}' OFS='/')`

The skopeo call to check if it exists or not is done per tag

mmalina replied:
Oh, so `ec track bundle` will push the $ACCEPTABLE_BUNDLES:$TAG image so that on the next tag it will already exist? I did not realize that. And given that most of the time there will be just one tag, as you say, it should be ok.

avi-biton replied:
Exactly @mmalina .
I also updated `skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep ${TAG} &> /dev/null`
to `skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep "^${TAG}$" &> /dev/null`
As you suggested

mmalina replied:
Thanks. I assumed the tag list would be a tag per line (I should have known better, given that you already had the `jq` there). So this won't work - see the new thread I created.

mmalina commented on tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml:
```diff
@@ -0,0 +1,91 @@
+            do
+                imageTag=$(jq -c -r --argjson j "$j" '.tags[$j]' <<< "$component")
+                
+                set +e
+                # Check if ACCEPTABLE_BUNDLES OCI artifact has a latest tag
+                skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}" | jq -r '.Tags[]' | grep "^${TAG}$" &> /dev/null
+                RESULT=$?
+                set -e
+
+                # If it has a latest tag, use it as an --input for the ec track bundle command
+                if [ $RESULT -eq 0 ]; then
```
I see you modified the grep based on my suggestion, but actually it won't work because the output of `skopeo list-tags` is a json, right? So something like this should work:

```suggestion
                set +e
                # Check if ACCEPTABLE_BUNDLES OCI artifact has a latest tag
                SKOPEO_TAGS=$(skopeo list-tags docker://"${ACCEPTABLE_BUNDLES}")
                RESULT=$?
                set -e

                # If it has a latest tag, use it as an --input for the ec track bundle command
                if [ "$RESULT" -eq 0 ] && [ -n "SKOPEO_TAGS" ] \
                  && [ "$(jq '.Tags | index(["'$TAG'"])' <<< "SKOPEO_TAGS")" != null ] ; then
```

avi-biton replied:
I actually tested it with an image from my repo (its public) and it seem to work:
```
TAG="sha256-d17e25a0afdc1a198d88ae303df1003ecd0dd2bf14f7d5378b2d4d7dfa4cca74"
$ skopeo list-tags docker://quay.io/avi_test/user-ns2/release-catalog/task-echo | jq -r '.Tags[]' | grep "${TAG}"
sha256-d17e25a0afdc1a198d88ae303df1003ecd0dd2bf14f7d5378b2d4d7dfa4cca74
sha256-d17e25a0afdc1a198d88ae303df1003ecd0dd2bf14f7d5378b2d4d7dfa4cca74.att
sha256-d17e25a0afdc1a198d88ae303df1003ecd0dd2bf14f7d5378b2d4d7dfa4cca74.sig

$ skopeo list-tags docker://quay.io/avi_test/user-ns2/release-catalog/task-echo | jq -r '.Tags[]' | grep "^${TAG}$"
sha256-d17e25a0afdc1a198d88ae303df1003ecd0dd2bf14f7d5378b2d4d7dfa4cca74
```

mmalina replied:
I think I'm tired today 😅 Thanks for the proof. So it should be fine now.

avi-biton commented:
@mmalina I am not sure if the e2e tests failures are related to my PR or not.
It seems to me that its not...

avi-biton commented:
/retest

mmalina commented:
> @mmalina I am not sure if the e2e tests failures are related to my PR or not. It seems to me that its not...

Actually, e2e is irrelevant here, since this is a new task and definitely not used in any e2e test. So I am going to merge this now.

avi-biton commented:
> > @mmalina I am not sure if the e2e tests failures are related to my PR or not. It seems to me that its not...
> 
> Actually, e2e is irrelevant here, since this is a new task and definitely not used in any e2e test. So I am going to merge this now.

Thanks a lot :)
